### PR TITLE
Add move task operation 

### PR DESCRIPTION
**Pull request by amitaekbote**
### Description of the Change

(edits by @fabiokung)

Implementation of the `MoveTask` operation, specified here: https://github.com/Netflix/titus-api-definitions/pull/75

Notable things missing on this first pass, and will come in future PRs:

1. Enhancements to the jobs/tasks event stream, signaling that a task has been moved to subscribers.
2. Validation that source and destination jobs are compatible (they have the same `Container` specs in their `JobDefinitions`.

The new API is behind a feature flag, and is disabled by default (for now).

### CODE REVIEW

File: titus-api/src/main/java/com/netflix/titus/api/jobmanager/service/JobManagerException.java [hunk 1]
```diff
@@ -45,6 +47,11 @@
         TaskTerminating,
         InvalidContainerResources,
         InvalidDesiredCapacity,
+        BelowMinCapacity,
```
[the remark this answers is not in the document]
Comment (tbak) [reply titus-api/src/main/java/com/netflix/titus/api/jobmanager/service/JobManagerException.java]: +1

File: titus-common/src/main/java/com/netflix/titus/common/framework/reconciler/ReconciliationFramework.java [hunk 2]
```diff
@@ -73,4 +74,12 @@
      * Removes an existing reconciliation engine.
      */
     Completable removeEngine(ReconciliationEngine<EVENT> engine);
+
+    /**
+     * Change reference entities for the provided list of the reconciliation engines. The returned observable completes
+     * successfully if the change action and the all model update actions completed successfully.
+     */
+    Observable<Void> changeReferenceModel(MultiEngineChangeAction multiEngineChangeAction,
```
[the remark this answers is not in the document]
Comment (tbak) [reply titus-common/src/main/java/com/netflix/titus/common/framework/reconciler/ReconciliationFramework.java]: We could do that, but in this case the change action is no longer a function, and I wanted to keep this property.
We could implement this as a higher order function, where in the first evaluation we deal wit the ids. Lets brainstorm this further.

File: titus-common/src/main/java/com/netflix/titus/common/framework/reconciler/internal/DefaultReconciliationFramework.java [hunk 3]
```diff
@@ -209,6 +216,54 @@ public Completable removeEngine(ReconciliationEngine engine) {
...
+
+        return Observable.create(emitter -> {
+
+            List<ReconciliationEngine<EVENT>> engines = new ArrayList<>();
+            for (String id : rootEntityHolderIds) {
+                ReconciliationEngine<EVENT> engine = findEngineByRootId(id).orElseThrow(() -> new IllegalArgumentException("Reconciliation engine not found: rootId=" + id));
+                engines.add(engine);
+            }
+
+            List<Observable<Map<String, List<ModelActionHolder>>>> outputs = ObservableExt.demultiplex(multiEngineChangeAction.apply(), engines.size());
```
[the remark this answers is not in the document]
Comment (tbak) [reply titus-common/src/main/java/com/netflix/titus/common/framework/reconciler/internal/DefaultReconciliationFramework.java]: I think this is documented in the interface.

File: titus-api/src/main/java/com/netflix/titus/api/jobmanager/service/JobManagerException.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (amitaekbote) [reply titus-api/src/main/java/com/netflix/titus/api/jobmanager/service/JobManagerException.java]: +1

File: titus-common/src/main/java/com/netflix/titus/common/util/rx/ObservableExt.java [hunk 4]
```diff
@@ -394,4 +394,17 @@ public static InstrumentedEventLoop createEventLoop(String metricNameRoot, Regis
...
+    /**
+     * Creates multiple {@link Observable} instances (outputs) backed by the same source observable, with the following properties:
+     * <ul>
+     * <li>the source observable is subscribed to only when the all outputs are subscribed to</li>
+     * <li>the source observable is subscribed to only once, and the same result is propagated to the all outputs</li>
+     * <li>if one of the output observables is cancelled, all the remaining observables are cancelled as well</li>
+     * <li>if an output is subscribed to multiple times, all the subscriptions except the first one are terminated with an error</li>
+     * </ul>
+     */
+    public static <T> List<Observable<T>> demultiplex(Observable<T> source, int outputs) {
```
[the remark this answers is not in the document]
Comment (amitaekbote) [reply titus-common/src/main/java/com/netflix/titus/common/util/rx/ObservableExt.java]: +1 for broadcast

File: titus-server-federation/src/main/java/com/netflix/titus/federation/service/AggregatingJobManagementClient.java [hunk 5]
```diff
@@ -366,6 +367,14 @@ public Completable killTask(TaskKillRequest request) {
         return result.toCompletable();
     }
 
+    @Override
+    public Completable moveTask(TaskMoveRequest taskMoveRequest) {
```
[the remark this answers is not in the document]
Comment (amitaekbote) [reply titus-server-federation/src/main/java/com/netflix/titus/federation/service/AggregatingJobManagementClient.java]: +1

File: titus-server-master/src/main/java/com/netflix/titus/master/jobmanager/service/DefaultV3JobOperations.java [hunk 6]
```diff
@@ -386,6 +388,67 @@ public Completable recordTaskPlacement(String taskId, Function<Task, Task> chang
...
+            ReconciliationEngine<JobManagerReconcilerEvent> engineFrom = fromEngineTaskPair.getLeft();
+            Job<ServiceJobExt> jobFrom = engineFrom.getReferenceView().getEntity();
+
+            // Validate the source job is a service job
+            if (!JobFunctions.isServiceJob(jobFrom)) {
+                throw JobManagerException.notServiceJob(jobFrom.getId());
+            }
+
+            // Validate that the task belongs to source job id
+            if(!jobFrom.getId().equals(sourceJobId)) {
```
[the remark this answers is not in the document]
Comment (amitaekbote) [reply titus-server-master/src/main/java/com/netflix/titus/master/jobmanager/service/DefaultV3JobOperations.java]: +1

File: titus-server-federation/src/main/java/com/netflix/titus/federation/service/AggregatingJobManagementClient.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (fabiokung) [reply titus-server-federation/src/main/java/com/netflix/titus/federation/service/AggregatingJobManagementClient.java]: After looking into it, the additional test won't bring much more value on top of the coverage we already have, so I will skip it.

File: titus-common/src/main/java/com/netflix/titus/common/util/rx/ObservableExt.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (fabiokung) [reply titus-common/src/main/java/com/netflix/titus/common/util/rx/ObservableExt.java]: I decided to go with `demultiplex -> propagate`.

File: titus-common/src/main/java/com/netflix/titus/common/util/rx/Demultipex.java [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+import com.netflix.titus.common.util.Evaluators;
+import com.netflix.titus.common.util.ExceptionExt;
+import rx.Emitter;
+import rx.Observable;
+import rx.Subscription;
+
+/**
+ * See {@link ObservableExt#demultiplex(Observable, int)} for more information.
+ */
+class Demultipex<T> {
```
[the remark this answers is not in the document]
Comment (fabiokung) [reply titus-common/src/main/java/com/netflix/titus/common/util/rx/Demultipex.java]: I renamed the operation to `ObservableExt.propagate()`, and the operator to `Propagator`. I think it captures the intent.

**Comment by fabiokung** [comment]
@amitaekbote @tbak @corindwyer I addressed the remaining comments, and I think this is ready for a final review.

File: titus-server-master/src/main/java/com/netflix/titus/master/jobmanager/service/DefaultV3JobOperations.java [hunk 8]
```diff
@@ -399,6 +401,69 @@ public Completable recordTaskPlacement(String taskId, Function<Task, Task> chang
...
+            if (!jobFrom.getId().equals(sourceJobId)) {
+                throw JobManagerException.taskJobMismatch(taskId, sourceJobId);
+            }
+
+            // Validate that we are moving across different jobs
+            if (jobFrom.getId().equals(targetJobId)) {
+                throw JobManagerException.sameJobs(jobFrom.getId());
+            }
+
+            // TODO(fabio): validate that source and destination jobs are compatible (same container specs)
```
Comment (corindwyer) [inline titus-server-master/src/main/java/com/netflix/titus/master/jobmanager/service/DefaultV3JobOperations.java]: Is this going to be in another PR?

File: titus-server-runtime/src/main/java/com/netflix/titus/runtime/endpoint/common/rest/TitusExceptionMapper.java [hunk 9]
```diff
@@ -226,9 +219,26 @@ private Response fromJobManagerException(JobManagerException e) {
             case TaskNotFound:
                 errorBuilder.status(HttpServletResponse.SC_NOT_FOUND);
                 break;
-            case NotServiceJob:
+            // these below are mapped to Status.FAILED_PRECONDITION for gRPC calls
+            case JobTerminating:
```
Comment (corindwyer) [inline titus-server-runtime/src/main/java/com/netflix/titus/runtime/endpoint/common/rest/TitusExceptionMapper.java]: We should not use internal server error for these. You can use status code 412 precondition failed.

File: titus-server-master/src/main/java/com/netflix/titus/master/jobmanager/service/DefaultV3JobOperations.java (hunk 8 above)
Comment (fabiokung) [reply titus-server-master/src/main/java/com/netflix/titus/master/jobmanager/service/DefaultV3JobOperations.java]: yes, please read the description of the PR :-)

File: titus-server-runtime/src/main/java/com/netflix/titus/runtime/endpoint/common/rest/TitusExceptionMapper.java (hunk 9 above)
Comment (fabiokung) [reply titus-server-runtime/src/main/java/com/netflix/titus/runtime/endpoint/common/rest/TitusExceptionMapper.java]: Will do. As discussed in person, the status code 412 is not really intended to be used for things like this (it is more related to headers passed in the request), but I think it is OK to put these errors into the 4xx category.
